Support subclass inheritance of pytree behavior

jekbradbury opened the pull request:
I'm exploring some ideas around object-oriented layers, where it would be helpful if subclasses could inherit pytree registration from their base classes. This PR makes pytree behavior inheritable by default, but also lets subclasses override any inherited behavior and allows registering a superclass in a non-inheritable way.

hawkinsp reviewed (approved):
Looks reasonable.

(A test might be nice, but not mandatory.)

mattjj commented:
It would be great if we could tweak this PR to solve #446. IIRC namedtuple instances are subclasses of `tuple` but have slightly different creation signatures, i.e. `TupleSubclass((1, 2, 3))` vs `NamedtupleSubclass(1, 2, 3)`, which might need handling.

jekbradbury committed:
Support subclass inheritance of pytree behavior, including namedtuples

This is somewhat complex when it comes to `tuple` and its subclasses: we 
want to support Python namedtuples, which are simply subclasses of tuple 
(albeit with a different constructor signature), but JAX core also uses 
tuple subclasses for values that are meant to be pytree leaves. In order 
to support this we add the ability to register a type as a pytree leaf, 
and register two core JAX types.

jekbradbury commented:
It's pretty unfortunate that Python namedtuples don't have an abstract base class--in our context, it makes them indistinguishable from core JAX types implemented as tuple subclasses (like PartialVal and JaxTuple) which need to be treated as leaves for the purposes of `pytree`. Rather than special-casing named tuples, I went ahead and implemented the ability to register a type as a leaf, and registering those two tuple subclasses was enough to get tests to pass.

I'm not sure if I should also add leaf registration for `AbstractTuple`, `Destructuring`, and `JaxprTracerTuple` (or any others)--I haven't looked into whether any of those can appear in the trees passed to the tree_utils machinery (leaving them out hasn't broken any tests).

jekbradbury commented:
(Also, adding a test was a great idea, because it turns out the code in the original PR didn't actually do anything).

levskaya commented:
I like the idea!  I agree that the way namedtuples are implemented in python is pretty annoying.  As an aside, you _can_ make namedtuple equivalents that have a (non-tuple) base class, this is one that should work equivalently in py2/py3 for all the ways we use them in JAX:
```python
from collections import Sequence

class SaneNamedTuple(Sequence): 
    """Subclass and define named fields with __slots__."""
    __slots__ = ()
    def __init__(self, *args):
        for slot, arg in zip(self.__slots__, args):
            setattr(self, slot, arg)
    def __repr__(self):
        attr_repr = ', '.join(map(lambda x: "%s=%r"%(x[0], x[1]),
                         tuple(zip(self.__slots__, tuple(self)))))
        return '{name}({args})'.format(name=type(self).__name__, args=attr_repr)
    def __iter__(self): 
        for name in self.__slots__:
            yield getattr(self, name)
    def __getitem__(self, index):
        return getattr(self, self.__slots__[index])
    def __len__(self):
        return len(self.__slots__)

def sane_namedtuple(cls_name, attr_names):
  "Create a named tuple-like class with name cls_name and fields attr_names."
  return type(cls_name,
              (SaneNamedTuple,), 
              {"__slots__": tuple(attr_names)})

Point = sane_namedtuple("Point", ("x", "y"))
pt = Point(1, 2)
print(pt)  # Point(x=1, y=2)
print(pt.x, pt.y)  # 1 2
print(isinstance(pt, Point))  # True
print(isinstance(pt, SaneNamedTuple))  # True
print(isinstance(pt, Sequence))  # True
print(isinstance(pt, tuple))  # False
x, y = pt
print(x, y)  # 1 2
```

jekbradbury committed:
add docstring

jekbradbury commented:
Added a docstring. Should be ready to merge if folks are okay with my evil reuse of the singleton intended to abstractly represent pytree leaves as a sentinel that represents pytree leaf type registrations.

hawkinsp commented:
(I think Matt would be the best reviewer for this, but he's on vacation so he might take a few days to get back to you.)

jekbradbury commented:
@levskaya the issue with that is the same limitation I'm running into trying to wrap stax layers in classes: it's not possible to create Python functions that expose a programmatically-defined argument signature (e.g. for tab completion or documentation generation) without using `exec`. Here that means that `SaneNamedTuple`s expose `*args` as their constructor signature rather than the set of fields in the named tuple.

levskaya commented:
@jekbradbury - I see what you want.  In python2 this is pretty much true. :(  In python3 we have the tools of inspect.Signature and inspect.Parameter which can set `__signature__` to do this though:
```python
     from inspect import Signature, Parameter
    @classmethod
    def get_signature(cls):
        parameters = [
          Parameter(name=slot, kind=Parameter.POSITIONAL_OR_KEYWORD) for slot in cls.__slots__
        ]
        return Signature(parameters=parameters)

    def __init_subclass__(cls, **kwargs):
        # ....
        cls.__slots__ = tuple(slots)
        cls.__signature__ = cls.get_signature()
```
yet another reason to drop py2 given it hasn't been improved in 10 years...

levskaya commented:
Sorry to keep distracting from the main point of this thread :)  too be concrete vis a vis above, this is the implementation of base-classable nametuple that I've seen before for py3.
```python
from collections import Sequence
from inspect import Signature, Parameter
class SaneNamedTuple(Sequence): 
    __slots__ = ()
    @classmethod
    def get_signature(cls):
        parameters = [Parameter(name=slot, kind=Parameter.POSITIONAL_OR_KEYWORD) 
                      for slot in cls.__slots__]
        return Signature(parameters=parameters)
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        slots = cls.__slots__
        cls.__slots__ = (tuple(slots.split()) if isinstance(slots, str) else tuple(slots))
        cls.__signature__ = cls.get_signature()
        cls.__init__.__signature__ = cls.get_signature()
        cls.__doc__ = ('{cls.__name__}{cls.__signature__}\n\n{cls.__doc__}'.format(cls=cls))
    def __new__(cls, *args, **kwargs):
        if cls is SaneNamedTuple:
            raise TypeError("Can't instantiate abstract class SaneNamedTuple")
        return super().__new__(cls)
    @classmethod
    def _get_bound_args(cls, args, kwargs):
        return Signature.bind(cls.__signature__, *args, **kwargs).arguments.items()
    def __init__(self, *args, **kwargs):
        bound_args = self._get_bound_args(args, kwargs)
        for slot, value in bound_args:
            setattr(self, slot, value)
    def __repr__(self):
        return type(self).__name__ + repr(tuple(self))
    def __iter__(self): 
        for name in self.__slots__:
            yield getattr(self, name)
    def __getitem__(self, index):
        return getattr(self, self.__slots__[index])
    def __len__(self):
        return len(self.__slots__)

def sane_namedtuple(cls_name, attr_names):
  "Create a named tuple-like class with name cls_name and fields attr_names."
  assert isinstance(attr_names, (list, tuple))
  return type(cls_name,
              (SaneNamedTuple,), 
              {"__slots__": tuple(attr_names)})
```

mattjj commented on jax/tree_util.py:
```diff
@@ -172,28 +171,66 @@ def __ne__(self, other):
+    str(py_type), to_iterable, from_iterable, include_subclasses)
 
-register_pytree_node(tuple, lambda xs: (xs, None), lambda _, xs: tuple(xs))
+def register_pytree_leaf(py_type):
```
Is `register_pytree_leaf` just optional?

It seems like it might be redundant: if we just have node types registered, then everything that isn't registered is a leaf. If we register both leaves and nodes, then in principle now unregistered things are a third thing (?), since they're neither a leaf nor a node.

I might be misunderstanding though. Can you explain the logic?

If indeed registering leaves is optional and does potentially break the current "registered = node, unregistered = leaf" dichotomy, then maybe we can remove it. What do you think?

mattjj commented on jax/tree_util.py:
```diff
@@ -172,28 +171,66 @@ def __ne__(self, other):
+      return node_type
+  return leaf
+
+def register_pytree_node(py_type, to_iterable, from_iterable,
```
Wonderful docstring, thank you!

mattjj commented on jax/tree_util.py:
```diff
@@ -172,28 +171,66 @@ def __ne__(self, other):
 
-def register_pytree_node(py_type, to_iterable, from_iterable):
+def get_node_type(py_type):
+  for cls in py_type.__mro__:
```
Accessing `pytype.__mro__` feels kind of dangerous, like we're going from something really simple and minimal to something that knows a lot about Python inheritance. Could this be fragile? Are there alternatives? For example, it seems like using `isinstance` on all the registered node types might be an alternative; I'm not saying that's better, but I'd like to think through the solution space.

I'm also interested if this has a big impact on performance. An advantage of the `type`-keyed dict mechanism is that it should be quite fast. I'm not saying the performance _is_ much different here, or that it's actually an important difference, but we should put some thought into that. What do you think?

mattjj commented on jax/tree_util.py:
```diff
@@ -172,28 +171,66 @@ def __ne__(self, other):
+  return leaf
+
+def register_pytree_node(py_type, to_iterable, from_iterable,
+                         include_subclasses=True):
```
Can we not have a default value here, or else leave the default to `False` so that previous behavior is maintained? In general I think default values can be bad and it's better to force the call site to be explicit.

mattjj commented on tests/core_test.py:
```diff
@@ -174,10 +174,13 @@ def foo(x):
-    ys = ({'a': 10}, [20, 30])
-    ys_bad = ({'a': 10, 'b': 10}, [20, 30])
-    zs = ({'a': 11}, [22, 33])
+    class MyDict(dict):
```
Instead of replacing the previous test, can you add a new one? It can just copy and paste things from the existing test. The more tests, the better!

mattjj reviewed (commented):
Thanks for this! And sorry for being slow to review.

I'm a little wary of this PR because it feels like it could be taking a very simple system and adding some nontrivial complexity. But I'm not sure about that because I don't have a good sense of how much complexity is actually here. (Surely the code in this PR is small and elegant, so maybe it's not actually much complexity after all!)

For that reason I left some questions generally about complexity (mostly conceptual and code complexity, but a bit about runtime computational complexity too). I'd really appreciate if you could help me think through the potential complexity here, so that we can decide the best course of action!

NeilGirdhar commented:
I've also been really interested in dealing with inheritance and pytrees.  I've been inheriting pytree-adaptation like this:

https://github.com/NeilGirdhar/efax/blob/master/efax/tensors/tensor_like.py

You can see how it's used here:

https://github.com/NeilGirdhar/efax/blob/630b73dadab01271f794119a6d41f5b6fce1bb66/efax/tensors/jax_generator.py#L34

I have been using it with inheritance too.  When there is no extra information, the subclass inherits the pytree conversion.  When there are extra members, you can override and call super:

    @classmethod
    def from_tensor(cls, values, aux_data):
        super_aux_data, distribution = aux_data
        r = super().from_tensor(values, super_aux_data)
        r.distribution = distribution
        return r

    def to_tensor(self):
        values, aux_data = super().to_tensor()
        return values, (aux_data, self.distribution)

jekbradbury commented:
I like this a lot! I was unaware of `__init_subclass__`; that's a pretty cool feature.

NeilGirdhar commented:
I thought I would throw up an issue since it would be nice to have https://github.com/google/jax/issues/2916

hawkinsp commented:
Do we still want to merge this or something along these lines, or should we close this PR?
